tokio/src/io/driver/scheduled_io.rs

```diff
@@ -57,51 +134,330 @@ impl ScheduledIo {
     /// generation, then the corresponding IO resource has been removed and
     /// replaced with a new resource. In that case, this method returns `Err`.
     /// Otherwise, this returns the previous readiness.
-    pub(crate) fn set_readiness(
+    pub(super) fn set_readiness(
         &self,
         token: Option<usize>,
+        tick: Tick,
         f: impl Fn(usize) -> usize,
     ) -> Result<usize, ()> {
         let mut current = self.readiness.load(Acquire);
 
         loop {
-            let current_generation = super::GENERATION.unpack(current);
+            let current_generation = GENERATION.unpack(current);
 
             if let Some(token) = token {
                 // Check that the generation for this access is still the
                 // current one.
-                if super::GENERATION.unpack(token) != current_generation {
+                if GENERATION.unpack(token) != current_generation {
                     return Err(());
                 }
             }
 
-            // Mask out the generation bits so that the modifying function
-            // doesn't see them.
+            // Mask out the tick/generation bits so that the modifying
+            // function doesn't see them.
             let current_readiness = current & mio::Ready::all().as_usize();
-            let new = f(current_readiness);
+            let mut new = f(current_readiness);
 
             debug_assert!(
-                new <= super::ADDRESS.max_value(),
-                "new readiness value would overwrite generation bits!"
+                new <= READINESS.max_value(),
+                "new readiness value would overwrite tick/generation bits!"
             );
 
-            match self.readiness.compare_exchange(
-                current,
-                super::GENERATION.pack(current_generation, new),
-                AcqRel,
-                Acquire,
-            ) {
+            match tick {
+                Tick::Set(t) => {
+                    new = TICK.pack(t as usize, new);
+                }
+                Tick::Clear(t) => {
+                    if TICK.unpack(current) as u8 != t {
+                        // Trying to clear readiness with an old event!
+                        return Err(());
+                    }
+                    new = TICK.pack(t as usize, new);
+                }
+            }
+
+            new = GENERATION.pack(current_generation, new);
+
+            match self
+                .readiness
+                .compare_exchange(current, new, AcqRel, Acquire)
+            {
                 Ok(_) => return Ok(current),
                 // we lost the race, retry!
                 Err(actual) => current = actual,
             }
         }
     }
+
+    pub(super) fn wake(&self, ready: mio::Ready) {
+        let mut waiters = self.waiters.lock().unwrap();
+
+        // check for AsyncRead slot
+        if !(ready & (!mio::Ready::writable())).is_empty() {
```

Comment:
out of  curiosity, why this logic rather than checking specifically for the presence of the `mio::Ready::readable` bit? might be worth a comment explaining that.

Reply:
I can't answer that, this was just taken from https://github.com/tokio-rs/tokio/pull/2828/files#diff-006d862c6a6ed63b5b19e9957059390fL185

Reply:
ah, never mind, then!